src/deephaven/plugin/object.py

```diff
@@ -17,46 +17,84 @@ def index(self) -> int:
         May be used in the output stream to refer to the reference from the client."""
         return self._index
 
-    @property
-    def type(self) -> Optional[str]:
-        """The type."""
-        return self._type
-
 
 class Exporter(abc.ABC):
-    """The interface for creating new references during ObjectType.to_bytes."""
+    """The interface for creating new references during FetchOnlyObjectBase.to_bytes."""
+
+    @abc.abstractmethod
+    def reference(self, obj: object) -> Reference:
+        """Creates a reference for an object. Each reference """
+        pass
+
+
+class MessageStream(abc.ABC):
+    """A stream of messages, either sent from server to client or client to server. ObjectType implementations
+    provide an instance of this interface for each incoming stream to invoke as messages arrive, and will
+    likewise be given an instance of this interface to be able to send messages to the client.
+    """
+    def __init__(self):
+        pass
 
     @abc.abstractmethod
-    def reference(
-        self, object, allow_unknown_type: bool = False, force_new: bool = False
-    ) -> Optional[Reference]:
-        """Gets the reference for object if it has already been created and force_new is False,
-        otherwise creates a new one. If allow_unknown_type is False, and no type can be found, no
-        reference will be created."""
+    def on_close(self):
+        """Closes the stream on both ends. No further messages can be sent or received."""
+        pass
+
+    @abc.abstractmethod
+    def on_data(self, payload: bytes, references: List[Any]):
+        """Transmits data to the remote end of the stream. This can consist of a binary payload and references to
+        objects on the server.
+        """
         pass
 
 
 class ObjectType(Plugin):
     """An object type plugin. Useful for serializing custom objects between the server / client."""
 
+    def __init__(self):
+        pass
+
     @property
     @abc.abstractmethod
     def name(self) -> str:
         """The name of the object type."""
         pass
 
     @abc.abstractmethod
-    def is_type(self, object) -> bool:
-        """Returns true if, and only if, the object is compatible with this object type."""
+    def is_type(self, obj: Any) -> bool:
+        """Returns True if, and only if, the object is compatible with this object type."""
         pass
 
+
+class BidirectionObjectType(ObjectType):
```

Comment:
Note that we could merge this into `ObjectType` directly, so that it is the default - the Java server currently tests if `create_client_connection` exists, and if not fall back to `to_bytes`, but we could reverse that - test for `to_bytes` first, and then fall back to the default of `create_client_connection`, if it exists. This would make it slightly easier to do the right thing. One additional change that would be required would be for `FetchOnlyObjectType` to override `create_client_connection`, so that the subclass didn't have an error for failing to override that method.

Reply:
@jmao-denver

Reply:
Probably should say Bidirectional anyway

Reply:
to prevent users from write the following code
```
from deephaven.plugin import object
```
we should rename object.py to object_type.py?

Reply:
from the PR description:
> Before merge, will also rename object.py to object_type.py, as long as we're doing other breaking changes, to avoid potential import confusion.